treeverse/lakeFS
# Introduce support of GCS encryption for both CMEK and CSEK

Closes #(Insert issue number closed by this PR)

## Change Description

### Background

Provide the support of GCS encryption for both CMEK and CSEK
      
### New Feature

Issue link: https://github.com/treeverse/lakeFS/issues/7557

- Configuration of CMEK supported
- Configuration of CSEK supported
- Throw an error for generating PreSignedURL of GCS with CSEK as the user must have the key in the configuration file
  - Frontend change is required
  - If user must have the key in configuration, then the CSEK make not much sense for PreSignedURL

### Testing Details

How were the changes tested?
* CMEK and CSEK cannot be configured at the same time, or the server will fail to start
* server will failed to start if CSEK is not a valid AES256 32bytes value
* CSEK encrypted object will fail to read when CSEK is not configured or wrong key
* CMEK encrypted object will fail to read when CMEK is not configured or wrong key
* Upload the file to a CMEK enabled bucket with CMEK configuration will be success
* Check the content of the CMEK encrypted file with CMEK configured
* Upload the file with CSEK will be success
* Check the content of the CSEK encrypted file with CSEK configured
* When the CSEK is configured and PreSignedURL is enabled, generating a PreSignedURL of a CSEK encrypted object is failed

### Breaking Change?

Does this change break any existing functionality? (API, CLI, Clients)

No breaking change as there's no API changed, only the server configuration is required

## Additional info

Logs, outputs, screenshots of changes if applicable (CLI / GUI changes)

### Contact Details

By GitHub  account
